app/routes_no_csrf/authentication-saml.js

```diff
@@ -3,21 +3,29 @@ const logger = require('../logger')
 
 module.exports = function (router) {
   router.get('/login', passport.authenticate('saml'), function (req, res) {
+    setRedirectTo(req.session, req.headers.referer)
     return res.redirect('/hmpps/0')
   })
 
   router.post('/login', passport.authenticate('saml'), function (req, res) {
     if (req.user) {
       logger.info(req.user.nameID, 'logged in.')
     }
-    return res.redirect('/')
+    return res.redirect(req.session.redirectTo || '/')
   })
 
   router.get('/logout', function (req, res) {
+    setRedirectTo(req.session, req.headers.referer)
     passport.logout(req, res)
     if (req.user) {
       logger.info(req.user.nameID, 'logged out.')
     }
     req.logout()
   })
 }
+
+var setRedirectTo = function (session, redirectTo) {
```

Comment:
Why do we need this function? We check the `redirectTo` variable is checked before read on line 14. If the `redirectTo` variable is undefined we will fallback to the default address.

Reply:
The function is used in the get request for logout and login routes. For the post route we don't don't need to do any set request call hence we only needed to check if the redirect is valid or not.

Reply:
Sorry, I didn't explain myself very well. I meant that we didn't need to worry about writing a `undefined` value to the session variable because we're handling it on read. However, thinking about it, it makes sense that if there is already a `redirectTo` field set we probably shouldn't overwrite it with undefined so happy for this to stay. Thanks.